bee-tangle/src/metadata.rs

```diff
@@ -0,0 +1,303 @@
+            milestone_index,
+            arrival_timestamp,
+            solidification_timestamp,
+            reference_timestamp,
+            omrsi,
+            ymrsi,
+            conflict,
+        }
+    }
+
+    /// Create metadata that corresponds to a just-arrived message using the current system time.
+    pub fn arrived() -> Self {
+        Self {
+            arrival_timestamp: SystemTime::now()
+                .duration_since(UNIX_EPOCH)
+                .expect("Clock may have gone backwards")
+                .as_millis() as u64,
+            ..Self::default()
+        }
+    }
+
+    /// Get the flags associated with this metadata.
+    pub fn flags(&self) -> &Flags {
+        &self.flags
+    }
+
+    /// Get a mutable reference to the flags associated with this metadata.
+    pub fn flags_mut(&mut self) -> &mut Flags {
+        &mut self.flags
+    }
+
+    /// Get the milestone index of this message.
+    pub fn milestone_index(&self) -> Option<MilestoneIndex> {
+        self.milestone_index
+    }
+
+    /// Set the milestone index of this message.
+    pub fn set_milestone_index(&mut self, index: MilestoneIndex) {
+        self.milestone_index = Some(index);
+    }
+
+    /// Get the arrival timestamp (seconds from the unix epoch) of this message.
+    pub fn arrival_timestamp(&self) -> u64 {
+        self.arrival_timestamp
+    }
+
+    /// Get the solidification timestamp (seconds from the unix epoch) of this message.
+    pub fn solidification_timestamp(&self) -> u64 {
+        self.solidification_timestamp
+    }
+
+    /// Get the oldest message root snapshot index of this message.
+    pub fn omrsi(&self) -> Option<IndexId> {
+        self.omrsi
+    }
+
+    /// Set the oldest message root snapshot index of this message.
+    pub fn set_omrsi(&mut self, omrsi: IndexId) {
+        self.omrsi = Some(omrsi);
+    }
+
+    /// Get the youngest message root snapshot index of this message.
+    pub fn ymrsi(&self) -> Option<IndexId> {
+        self.ymrsi
+    }
+
+    /// Set the youngest message root snapshot index of this message.
+    pub fn set_ymrsi(&mut self, ymrsi: IndexId) {
+        self.ymrsi = Some(ymrsi);
+    }
+
+    /// Get the reference timestamp (seconds from the unix epoch) of this message.
+    pub fn reference_timestamp(&self) -> u64 {
+        self.reference_timestamp
+    }
+
+    /// Solidify this message at the current system time.
+    pub fn solidify(&mut self) {
+        self.flags.set_solid(true);
+        self.solidification_timestamp = SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .expect("Clock may have gone backwards")
+            .as_millis() as u64;
+    }
+
+    /// Reference this message with the given timestamp.
+    pub fn reference(&mut self, timestamp: u64) {
+        self.flags.set_referenced(true);
+        self.reference_timestamp = timestamp;
+    }
+
+    /// Get the conflict state of this message.
+    pub fn conflict(&self) -> u8 {
```

Comment:
I'm not sure. @thibault-martinez do you have opinions on this?

Reply:
There is actually a type in the ledger crate but we couldn't use it because of a circular dependency. Since, the ledger types have been put in a module that doesn't use the tangle so maybe we can start using it ? It's called `ConflictReason`